### PR TITLE
wr-223: Add ability to follow specific author

### PR DESCRIPTION
**Pull request by drMangust**
image

### CODE REVIEW

File: backend/src/db/migrations/20230915125423_create_user_following_authors_table.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+const TableName = {
+  USERS: 'users',
+  USER_FOLLOWING_AUTHORS: 'user_following_authors',
+} as const;
+
+const ColumnName = {
+  ID: 'id',
+  USER_ID: 'user_id',
+  AUTHOR_ID: 'author_id',
+  GENRE_ID: 'genre_id',
```
Comment (andanf-e) [inline backend/src/db/migrations/20230915125423_create_user_following_authors_table.ts]:
```suggestion
```

File: backend/src/packages/follow/follow.repository.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        .castTo<{ count: string }>();
+
+      const isFollowed = userId
+        ? await this.followModel.query(trx).where({ userId, authorId }).first()
+        : false;
+
+      return {
+        authorId,
+        followersCount: Number.parseInt(followersCount?.count, 10),
+        isFollowed: !!isFollowed,
```
Comment (andanf-e) [inline backend/src/packages/follow/follow.repository.ts]:
```suggestion
        isFollowed: Boolean(isFollowed),
```

File: shared/src/packages/users/libs/enums/users-api-path.enum.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 const UsersApiPath = {
   ROOT: '/',
+  FOLLOW: '/:id/follow',
```
Comment (andanf-e) [inline shared/src/packages/users/libs/enums/users-api-path.enum.ts]: To keep it consistent with value
```suggestion
  $ID_FOLLOW: '/:id/follow',
```

File: frontend/src/pages/article/shared-article-page.tsx [hunk 4]
```diff
@@ -62,6 +62,8 @@ const SharedArticlePage: React.FC = () => {
                   publishedAt={article.publishedAt ?? ''}
                   genre={article.genre ?? ''}
                   avatarUrl={article.author.avatarUrl}
+                  authorFollowers={article.author.followersCount}
+                  isShared={true}
```
Comment (andanf-e) [inline frontend/src/pages/article/shared-article-page.tsx]: Could be simplified
```suggestion
                  isShared
```

File: backend/src/db/migrations/20230915125423_create_user_following_authors_table.ts (hunk 1 above)
Comment (drMangust) [reply backend/src/db/migrations/20230915125423_create_user_following_authors_table.ts]: [5ec0e7e](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/5ec0e7eec42503afe6d1938876c98865354af9ca)

File: backend/src/packages/follow/follow.repository.ts (hunk 2 above)
Comment (drMangust) [reply backend/src/packages/follow/follow.repository.ts]: [5522c52](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/5522c5224d09b95fd29abdab3b684aa0c39b547c)

File: frontend/src/pages/article/shared-article-page.tsx (hunk 4 above)
Comment (drMangust) [reply frontend/src/pages/article/shared-article-page.tsx]: [1138609](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/1138609f69c165918af13092d761523d0baed61c)

File: shared/src/packages/users/libs/enums/users-api-path.enum.ts (hunk 3 above)
Comment (drMangust) [reply shared/src/packages/users/libs/enums/users-api-path.enum.ts]: [5509f38](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/5509f38433b30fe7a12f4abd7441e6c40deda542)

File: backend/src/packages/follow/follow.repository.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        .first()
+        .castTo<{ count: string }>();
+
+      const isFollowed = userId
+        ? await this.followModel.query(trx).where({ userId, authorId }).first()
+        : false;
+
+      return {
+        authorId,
+        followersCount: Number.parseInt(followersCount?.count, 10),
```
Comment (andanf-e) [inline backend/src/packages/follow/follow.repository.ts]: Can we cast it this way?
```suggestion
        followersCount: Number(followersCount?.count ?? 0),
```

Comment (drMangust) [reply backend/src/packages/follow/follow.repository.ts]: @andanf-e 
We can do it this way:
it should return 0 anyway if there is nothing
```suggestion
        followersCount: Number(followersCount.count),
```

Comment (drMangust) [reply backend/src/packages/follow/follow.repository.ts]: [c376cc8](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/c376cc8dc18ae8989f577807fafaada40071ae37)

**Review by andanf-e — DISMISSED** [review]

File: frontend/src/pages/article/components/article-details/article-details.tsx [hunk 6]
```diff
@@ -49,6 +57,13 @@ const ArticleDetails: FC<Properties> = ({
             rating
           </li>
         </ul>
+        {!isArticleOwner && !isShared && (
+          <Button
+            className={styles.followButton}
+            label={isFollowed ? 'Unfollow' : 'Follow'}
+            onClick={onFollow}
```
Comment (hnoievets) [inline frontend/src/pages/article/components/article-details/article-details.tsx:64]: Maybe we should put 'Unfollow' and 'Follow' in an enum?

Comment (drMangust) [reply frontend/src/pages/article/components/article-details/article-details.tsx:64]: sounds very logical

Comment (drMangust) [reply frontend/src/pages/article/components/article-details/article-details.tsx:64]: [979f92c](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/280/commits/979f92c0d9db98b3d536939a605cbd07d58e31f4)

**Review by andanf-e — APPROVED** [review]

**Review by NazarKovtun1 — APPROVED** [review]

**Pull request merged**